BSP/sharpMemoryLCD/LS013B7DH03.c

```diff
@@ -0,0 +1,255 @@
+    SpiHandle.Init.NSS               = SPI_NSS_SOFT;
+    SpiHandle.Init.NSSPMode          = SPI_NSS_PULSE_DISABLE;
+    SpiHandle.Init.Mode              = SPI_MODE_MASTER;
+    if(HAL_SPI_Init(&SpiHandle) != HAL_OK)
+    {
+        returnSuccess= false;
+    }
+
+    returnSuccess&= sharpMemoryLCD_clearScreen();
+
+    LCDdisplayEnable(true);
+
+    return returnSuccess;
+}
+
+
+bool sharpMemoryLCD_clearScreen(void)
+{
+    bool returnSuccess= true;
+    /* To clear sreen send CMD_CLEAR (3bit) + 16 additional dummy bits */
+    uint8_t buffer[] = {CMD_CLEAR, 0x00};
+    LCDslaveSelect(true);
+    if(HAL_SPI_Transmit(&SpiHandle, buffer, sizeof(buffer), 1000) != HAL_OK)
+    {
+        returnSuccess= false;
+    }
+    LCDslaveSelect(false);
+
+    return returnSuccess;
+}
+
+
+bool sharpMemoryLCD_printTextLine(uint8_t line, const char *text, uint8_t nbChar)
+{
+    bool returnSuccess = true;
+    uint16_t nbBytePixelAsciiLine= Font16.Height*SCREEN_LENGTH/Font16.Width;
+    uint8_t pixelBuf[nbBytePixelAsciiLine];
+    memset(pixelBuf, 0, nbBytePixelAsciiLine);
+
+    if(line < (SCREEN_HEIGHT/Font16.Height))
+    {
+        uint16_t screenX=0;
+        uint16_t XbytesPerFontChar = (Font16.Width+7)/8;
+        uint16_t bytesPerFontChar = XbytesPerFontChar*Font16.Height;
+        //end display at first character out of ascii (most common will be \0) or out of screen
+        for(uint8_t c=0; (c<nbChar) && (' '<=text[c]) && (text[c]<='~') && (c<(SCREEN_LENGTH/Font16.Width)); c++)
+        {
+            // here SIMD instructions could be used to process 4 lines in one instruction
+            for(uint8_t x=0; x<Font16.Width; x++)
+            {
+                for(uint8_t y=0; y<Font16.Height; y++)
+                {
+                    // search for the byte in font table where to find the info of the required pixel
+                    // the font table is a 1D array but it could be 3D
+                    // 1D is for the current ASCII char in the table
+                    // 1D is for the line inside the ASCII char description
+                    // 1D is for the byte inside the pixel line of the the ASCII char description
+                    uint16_t currentByteInFont = bytesPerFontChar*(text[c]-' ') + XbytesPerFontChar*y + x/8;
+                    // TODO could be optimized to do several pixel at one time to go until byte alignement
+                    // we need to update pixels inside a byte.
+                    // modify the pixel/bit in the correct byte of the buffer
+                    // by taking the correct bit in the correct byte of font table
+                    pixelBuf[y*SCREEN_LENGTH + screenX/8] |= (1<<(screenX % 8) & Font16.table[currentByteInFont]);
+                }
+                screenX++;
+            }
+        }
+        returnSuccess &= LCDupdateDisplay(line*Font16.Height + 1, pixelBuf, sizeof(pixelBuf));
+    }
+    else
+    {
+        returnSuccess = false;
```

Comment:
You could also start writing till the end of the screen in this case :)

Reply:
This one is an ascii line not a pixel line so I made it fail to not having half characters printed